packages/tinybased/src/lib/queries/QueryBuilder.ts

```diff
@@ -0,0 +1,308 @@
+  TStartTable extends OnlyStringKeys<TSchema> = never,
+  TJoinedTables extends keyof TSchema = never,
+  TSelection extends Record<string, unknown> = {},
+  TResult extends Record<string, unknown> = {}
+> {
+  private readonly joins: Array<[string, string]> = [];
+  private readonly selects: string[] = [];
+  private readonly wheres: Array<[string, Cell]> = [];
+  private readonly selectsWithAlias: Array<[string, string]> = [];
+  private readonly selectFroms: Array<[string, string]> = [];
+  private readonly selectFromsWithAlias: Array<[string, string, string]> = [];
+  private readonly whereFroms: Array<
+    [string, string, string | number | boolean]
+  > = [];
+  private readonly groups: Array<[string, Aggregations, string]> = [];
+  private readonly groupUsings: Array<[string, (any: []) => number, string]> =
+    [];
+
+  constructor(
+    private readonly queries: Queries,
+    private readonly startTable: OnlyStringKeys<TSchema>,
+    private readonly relationshipDefs: RelationshipDefinition[]
+  ) {}
+
+  join<
+    TRelationshipName extends MatchingRelationships<TStartTable, TRelationships>
+  >(
+    _relationshipName: TRelationshipName
+  ): QueryBuilder<
+    TSchema,
+    TRelationships,
+    TStartTable,
+    TJoinedTables | TRelationships[TRelationshipName]['to'],
+    TSelection,
+    TResult
+  > {
+    const { to, cell } = this.relationshipDefs.find(
+      (r) => r.name === _relationshipName
+    )!;
+    this.joins.push([to, cell]);
+    return this;
+  }
+
+  /**
+   * Selects a cell from the starting table of the query
+   */
+  select<TCellName extends OnlyStringKeys<TSchema[TStartTable]>>(
+    cellName: TCellName
+  ): QueryBuilder<
+    TSchema,
+    TRelationships,
+    TStartTable,
+    TJoinedTables,
+    TSelection & Record<TCellName, TSchema[TStartTable][TCellName]>,
+    TResult & Record<TCellName, TSchema[TStartTable][TCellName]>
+  > {
+    this.selects.push(cellName);
+    return this as any;
+  }
+
+  /**
+   * Selects a cell from the starting table and provides an alias for it
+   */
+  selectAs<
+    TCellName extends OnlyStringKeys<TSchema[TStartTable]>,
+    TAlias extends string
+  >(
+    cellName: TCellName,
+    alias: TAlias
+  ): QueryBuilder<
+    TSchema,
+    TRelationships,
+    TStartTable,
+    TJoinedTables,
+    TSelection & Record<TAlias, TSchema[TStartTable][TCellName]>,
+    TResult & Record<TAlias, TSchema[TStartTable][TCellName]>
+  > {
+    this.selectsWithAlias.push([cellName, alias]);
+    return this as any;
+  }
+
+  /**
+   * Creates an aggreation on one of the selected fields
+   */
+  group<TGroupCell extends OnlyStringKeys<TSelection>, TAlias extends string>(
+    groupBy: TGroupCell,
+    agg: Aggregations,
+    alias: TAlias
+  ): QueryBuilder<
+    TSchema,
+    TRelationships,
+    TStartTable,
+    TJoinedTables,
+    TSelection,
+    Record<TAlias, number>
+  > {
+    this.groups.push([groupBy, agg, alias]);
+    return this as any;
```

Comment:
This way may be we able to get rid of anys
```suggestion
    alias: TAlias
  ) {
    this.groups.push([groupBy, agg, alias]);
    return this as QueryBuilder<
      TSchema,
      TRelationships,
      TStartTable,
      TJoinedTables,
      TSelection,
      Record<TAlias, number>
    >;
```

Reply:
When the fluent APIs I prefer if we can be explicit on the signature and then have the inner code actually match that without casting. I'd rather figure out why the addition of the `build` method made it so that methods returning `this` no longer matches the signature